virtool/oidc/utils.py

```diff
@@ -0,0 +1,65 @@
+import json
+import os
+from dataclasses import dataclass, asdict
+
+import jwt
+from aiohttp.web import Application
+from jose.jwt import decode, get_unverified_header
+
+
+@dataclass
+class JWKArgs:
+    kty: str
+    kid: str
+    use: str
+    n: str
+    e: str
+
+
+async def validate_token(token: jwt, app: Application) -> dict:
+    """
+    Validate token and return claims using JWK_ARGS environment variable.
+
+    :param token: JWT token to validate
+    :param app: the application object
+    :return: JWT claims
+    """
+    try:
+        jwk_args = app["jwk_args"]
+    except KeyError:
+        jwk_args = await update_jwk_args(token, app)
+
+    return decode(
+        token,
+        asdict(jwk_args),
+        algorithms=["RS256"],
+        audience=os.environ.get("B2C_CLIENT_ID")
+    )
+
+
+async def update_jwk_args(token: jwt, app: Application) -> JWKArgs:
+    """
+    Update jwk_args and store in environment variable
+
+    :param token: JWT token to validate and decode
+    :param app: the app object
+    :return: jwk_args
+    """
+    header = await app["run_in_thread"](get_unverified_header, token)
+    authority = os.environ.get("B2C_AUTHORITY")
+    resp = await app["client"].get(f"{authority}/discovery/v2.0/keys", allow_redirects=True)
+    jwks_keys = json.loads(await resp.text())
+
+    jwk = [key for key in jwks_keys["keys"] if key["kid"] == header["kid"]][0]
```

Comment:
This is not very efficient.

You are iterating through the entire sequence even if you get a `kid` match in the first item. Move code into separate function that iterates through `jwks_keys` and returns the first matching key.

Also, is it supposed to be `jwk_keys`?

Reply:
I'll move to a separate function. JWKS stands for Json Web Key Set so I think I should just replace with `jwks` alone.